packages/api/bin/serveUtils.js

```diff
@@ -1,72 +1,197 @@
+      const dbRecord = await translateApiGranuleToPostgresGranule(dynamoRecord, knex);
+
+      const executionPgModel = new ExecutionPgModel();
+      const executionCumulusId = await executionPgModel.getRecordCumulusId(knex, {
+        url: g.execution,
+      });
+
+      await upsertGranuleWithExecutionJoinRecord(knex, dbRecord, executionCumulusId);
+    })
   );
 }
 
 async function addProviders(providers) {
+  const knex = await getKnexClient({
+    env: {
+      ...envParams,
+      ...localStackConnectionEnv,
+    },
+  });
+
   const providerModel = new models.Provider();
   const es = await getESClientAndIndex();
+  const providerPgModel = new ProviderPgModel();
   return await Promise.all(
-    providers.map((p) =>
-      providerModel
-        .create(p)
-        .then((provider) => indexer.indexProvider(es.client, provider, es.index)))
+    providers.map(async (p) => {
+      const dynamoRecord = await providerModel.create(p);
+      await indexer.indexProvider(es.client, dynamoRecord, es.index);
+      const dbRecord = await translateApiProviderToPostgresProvider(dynamoRecord);
+      await providerPgModel.create(knex, dbRecord);
+    })
   );
 }
 
 async function addRules(rules) {
+  const knex = await getKnexClient({
+    env: {
+      ...envParams,
+      ...localStackConnectionEnv,
+    },
+  });
+
   const ruleModel = new models.Rule();
   const es = await getESClientAndIndex();
+  const rulePgModel = new RulePgModel();
   return await Promise.all(
-    rules.map((r) =>
-      ruleModel
-        .create(r)
-        .then((rule) => indexer.indexRule(es.client, rule, es.index)))
+    rules.map(async (r) => {
+      const dynamoRecord = await ruleModel.create(r);
+      await indexer.indexRule(es.client, dynamoRecord, es.index);
+      const dbRecord = await translateApiRuleToPostgresRule(dynamoRecord, knex);
+      await rulePgModel.create(knex, dbRecord);
+    })
   );
 }
 
 async function addExecutions(executions) {
+  const knex = await getKnexClient({
+    env: {
+      ...envParams,
+      ...localStackConnectionEnv,
+    },
+  });
+
   const executionModel = new models.Execution();
   const es = await getESClientAndIndex();
-  return await Promise.all(
-    executions.map((e) =>
-      executionModel
-        .create(e)
-        .then((execution) =>
-          indexer.indexExecution(es.client, execution, es.index)))
-  );
+
+  executions.sort((firstEl, secondEl) => {
+    if (!firstEl.parentArn && !secondEl.parentArn) {
+      return 0;
+    }
+
+    if ((!firstEl.parentArn && secondEl.parentArn) || firstEl.arn === secondEl.parentArn) {
+      return -1;
+    }
+
+    return 1;
+  });
+
+  // Since executions has a parent/child relationship with itself
+  // a fake promise is used with reduce() to force records to be created
+  // synchronously
+  const executionPgModel = new ExecutionPgModel();
+  const starterPromise = Promise.resolve(null);
+  return await executions.reduce((p, e) => p
```

Comment:
Is the whole intent of this logic just to make sure that we create all the records serially rather than in parallel to ensure we don't get creation errors?

If so, this is a very clever solution, but I think it would be cleaner to use [`p-each-series`](There is no need to `omit` if we're comparing against the fetched record from dynamo which has the timestamps), which we already have as a dependency of the `api`

Reply:
Right. That is the only purpose for that logic. Interesting. i've never used p-each-series. Looking at it now.

Reply:
Wow. That's much cleaner. Updated and pushed.